ink-wrapper/src/codegen.rs

```diff
@@ -387,44 +385,40 @@ fn define_event(event: &EventSpec<PortableForm>, metadata: &InkProject) -> rust:
         $(event.label()) {
             $(for field in event.args() {
                 $(docs(field.docs()))
-                $(field.label()): $(event_field_type(field, metadata)),
+                $(field.label()): $(type_ref_prefix(field.ty().ty().id(), metadata, "super::")),
             })
         },
 
         $[ '\n' ]
     }
 }
 
-/// Helper function to generate the type of an event field.
-fn event_field_type(field: &EventParamSpec<PortableForm>, metadata: &InkProject) -> rust::Tokens {
-    let type_id = field.ty().ty().id();
-    let type_ref = type_ref(type_id, metadata);
-
-    if resolve(metadata, type_id).is_custom() {
-        quote! { super::$(type_ref) }
-    } else {
-        quote! { $(type_ref) }
-    }
+/// Generates a type reference to the given type (for example to use as an argument type, return type, etc.).
+fn type_ref(id: u32, metadata: &InkProject) -> String {
+    type_ref_prefix(id, metadata, "")
 }
 
 /// Generates a type reference to the given type (for example to use as an argument type, return type, etc.).
-fn type_ref(id: u32, metadata: &InkProject) -> String {
+///
+/// The `prefix` is prepended to the type name if the type is a custom type.
+fn type_ref_prefix(id: u32, metadata: &InkProject, prefix: &str) -> String {
     let typ = resolve(metadata, id);
+    let generic_prefix = if typ.is_custom() { prefix } else { "" };
```

Comment:
I wonder - shouldn't the caller of `type_ref_prefix` know whether it's needed or not? i.e., what's the situation when caller passes in _some_ prefix (not `""`) but `type.is_custom() == false`? What's the outcome of `[u64; 4].is_custom()`? Or `[MyStruct; 4].is_custom()`?

Reply:
No one likes the `Reader` monad. Even its mother.